plugins/uefi/fu-uefi-bootmgr.c

```diff
@@ -273,14 +274,23 @@ fu_uefi_bootmgr_bootnext (const gchar *esp_path,
 	secure_boot = fu_efivar_secure_boot_enabled ();
 	if (secure_boot) {
 		/* test to make sure shim is there if we need it */
-		shim_app = fu_uefi_get_esp_app_path (esp_path, "shim", error);
+		shim_app = fu_uefi_get_esp_app_path (device, esp_path, "shim", error);
 		if (shim_app == NULL)
 			return FALSE;
 
+		/* try to fallback to use UEFI removable path if the shim path doesn't exist */
+		if (!g_file_test (shim_app, G_FILE_TEST_EXISTS)) {
+			if (fu_device_get_metadata_boolean (device, "FallbacktoRemovablePath")) {
```

Comment:
maybe this is a stupid question, but why do we need FallbacktoRemovablePath at all? I'd really like to avoid adding more configuration, especially when we want to use the same snap image for Ubuntu Core and real Ubuntu.

Reply:
It's called the "Removable" path because it's intended for when booting removable media.  Abusing it for HDD/SSD on regular boot may be problematic based upon the UEFI implementation and how IBV/OEM choose to use it in their BDS implementation.  If we automatically fall back to this path all the time we're encouraging this behavior.

The only reason it's been used in Linux ecosystem is optional and for when there are bugs in UEFI implementation.  See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=767037 for an example.

Reply:
> especially when we want to use the same snap image for Ubuntu Core and real Ubuntu.

Regarding this point I would expect that the confined snap would have this option set to `true` and **nowhere** else.  If the confined snap gets used outside of Ubuntu core regular paths get used and on Ubuntu core this thing gets used.

Reply:
>  Abusing it for HDD/SSD on regular boot may be problematic

Can we push back on Ubuntu Core to stop doing that then?

Reply:
I cited from our production engineering to express some points of this design: It allows booting as removable media (cdrom / sd-card / usb stick) but also as internal drive (i.e. flashed in factory) without any modifications of an EFI variables. Also it is intentional that neither fallback, .CSV, nor mock-manager are provided out of the box. As using any of those features will affect the PCR measurements against which FDE secret is sealed against. The miminise number of possible bootflows and PCR values, it is intentional that UC20 provides more rigid structure (and less featureful ESP) when compared to classic. The things we chose to not ship would mostly open up ways for users to brick their device instead of achieving similar results as on classic.

Reply:
> It allows booting as removable media (cdrom / sd-card / usb stick) but also as internal drive (i.e. flashed in factory) 

By doing it this way, you are not exercising efi variables ever and if there _is_ a UEFI implementation BDS bug you won't encounter it until you tried to do a firmware update.  At which point you _might_ not be able to boot anymore.

>As using any of those features will affect the PCR measurements against which FDE secret is sealed against

I don't think it's a reasonable expectation that booting external media and booting internal media will have the same PCR values.  PCR4 will change based upon which entry is selected, and you no longer have a predictable entry.  Furthermore, you have no guarantee taking an external media between machines that all the PCRs will measure the same between machines.  If one has a dbx update installed and the other doesn't for example.

Furthermore, I don't think it makes sense to have FDE on an external media that you don't have a boot entry for.  You'll need to pick that entry from the boot list, or modify BootOrder to have that entry.  And now if you have multiple disks plugged in, or switch the port it's plugged into you break things!

> without any modifications of an EFI variables

It's impossible to do fwupd based updates unless efi variables are modified.

> Also it is intentional that neither fallback, .CSV, nor mock-manager are provided out of the box

This makes sense to me, but it doesn't mean you have to install the bootloader into /efi/boot.  It can still go into `/efi/ubuntu-core` and make an NVRAM entry.

>flashed in factory

Flashed "same" image in the factory doesn't need to mean just `dd if=foo.bin of=/dev/nvmen0`.  All reasonable factory processes offer the ability to write to NVRAM variables as well.  There are tools and APIs for Windows, and obviously Linux has `efibootmgr`.

Reply:
Maybe I'm put in the difficult position between our design and your thoughts. If this PR can't be acceptable, the only thing I could do is to use `EFI_OS_DIR`, and unfortunately I will keep maintaining patches. Before that, let me discuss internally.

Reply:
From linked bug recent discussion it sounds like there may be some appetite to using NVRAM variables instead.  Let's wait until conclusion there to revisit.  The actual code in this looks OK to me if this is the way we end up all agreeing to go.

Reply:
unfortunately this is time criticial. We need this PR as is, even if in the future nvram variable writing comes in. I fear the nvram writing might only happen in 2021 if not 2022.

Reply:
> I fear the nvram writing might only happen in 2021 if not 2022

Would the nvram writing ever happen if we merged this PR?

Reply:
> > I fear the nvram writing might only happen in 2021 if not 2022
> 
> Would the nvram writing ever happen if we merged this PR?

Yeah. Definatly. We still need bootentries to selftest / factory mode / etc. but backlog and priorities.

Reply:
When this would happen in 2021/2022, would older systems also eventually be covered?  I think we would like to be able to remove this option from fwupd at some point to avoid incurring technical debt.

Reply:
even if we add the bootentires, we will continue to use the UEFI removable path on disk. this is to stay long term.

Reply:
So if you still keep using the UEFI removable path but add NVRAM entries that's another interesting failure you might cause.  There are UEFI BDS implementations that will do "de-duplication" of entries.  Depending upon their algorithm either the entry for UEFI removable path or the entry you added might get removed.

Really the safest thing to do is to follow what the rest of the OS ecosystem does and create your own path in the ESP, place your EFI applications there and create NVRAM entries.  Otherwise be prepared for the wild west of firmware behaviors.

@hughsie Since even if ubuntu core adds nvram options they'll keep using UEFI removable path from an upstream fwupd perspective I think from what's been discussed there are these options:
1. We take this patch as is today, but keep upstream default as disabled, it's only used in their confined snap.
2. We add UEFI removable path as a permanent fallback.  This might have interesting implications for live media that we should think about (_maybe_ able to install UEFI BIOS updates from live media if your media is FAT32 and not written as `iso9660` (from a `dd`))
3. We reject this and Ubuntu core carries (or hardcodes) the fallback in confined snap as a downstream patch.

Reply:
@superm1 I trust your judgement on this. I'm not super keen on new config options but also see the benefit of having the same codebase for both snaps.